Raven.CppClient/DocumentStore.h

```
@@ -11,37 +16,49 @@ namespace ravendb::client::documents
 		mutable std::map<std::string, std::shared_ptr<http::RequestExecutor>, impl::utils::CompareStringsIgnoreCase> _request_executors{};
 		mutable std::shared_mutex _request_executors_mutex{};
 
-		std::shared_ptr<operations::MaintenanceOperationExecutor> _maintenance_operation_executor = 
-			std::make_shared<operations::MaintenanceOperationExecutor>(*this);
-		std::shared_ptr<operations::OperationExecutor> _operation_executor =
-			std::make_shared<operations::OperationExecutor>(*this);
+		std::unique_ptr<identity::MultiDatabaseHiLoIdGenerator>_multi_db_hilo{};
+
+		std::shared_ptr<operations::MaintenanceOperationExecutor> _maintenance_operation_executor{};	
+		std::shared_ptr<operations::OperationExecutor> _operation_executor{};			
 
 		std::string _identifier{};
 
+		std::weak_ptr<DocumentStore> _weak_this;
+
+		void initiate_operation_executors();
+
+		DocumentStore();
+
+		DocumentStore(std::string url, std::string database);
+
+		DocumentStore(std::vector<std::string> urls, std::string database);
+
 	protected:
 		void assert_valid_configuration() const;
 
 	public:
 		~DocumentStore() override;
 
-		DocumentStore();
+		static std::shared_ptr<DocumentStore> create();
 
-		DocumentStore(std::string url, std::string database);
+		static std::shared_ptr<DocumentStore> create(std::string url, std::string database);
 
-		DocumentStore(std::vector<std::string> urls, std::string database);
+		static std::shared_ptr<DocumentStore> create(std::vector<std::string> urls, std::string database);
 
 		std::string get_identifier() const override;
 
 		void set_identifier(std::string identifier) override;
 
+		void close();
```

Why do we need something separate from the destructor?

---

Because `close()` might throw and we want to avoid exceptions thrown from a dtor.

---

+1 on not throwing from dtor.
But what errors could cause the close of the document store to throw?

---

To start with, the Store on `close()` returns the unused HiLo ranges to the server ...

---

Failing to return the unused hilo isn't an error. It is a courtesy, any errors doing that should be silently ignored.

---

`close()` is call from within the `~DocumentStore()` ignoring exceptions.

---

Why have it in the first place in the interface, then?

---

In Java client `close()`  is part of the `Closeable` interface.
Here I left it `public` so the user can close the store manually and treat
all would-be-thrown exceptions before actually freeing the object's memory.